Mathlib/LinearAlgebra/RootSystem/Basic.lean

```diff
@@ -41,182 +44,48 @@ variable {ι R M N : Type*}
 
 namespace RootPairing
 
-variable (P : RootPairing ι R M N) (i j : ι)
+section reflection_perm
 
-lemma root_ne (h: i ≠ j) : P.root i ≠ P.root j := by
-  simp_all only [ne_eq, EmbeddingLike.apply_eq_iff_eq, not_false_eq_true]
+variable (p : PerfectPairing R M N) (root : ι ↪ M) (coroot : ι ↪ N) (i j : ι)
+  (h : ∀ i, MapsTo (preReflection (root i) (p.toLin.flip (coroot i))) (range root) (range root))
 
-lemma ne_zero [CharZero R] : (P.root i : M) ≠ 0 :=
-  fun h ↦ by simpa [h] using P.root_coroot_two i
+private theorem exist_eq_reflection_of_mapsTo  :
+    ∃ k, root k = (preReflection (root i) (p.toLin.flip (coroot i))) (root j) :=
+  h i (mem_range_self j)
 
-lemma ne_zero' [CharZero R] : (P.coroot i : N) ≠ 0 :=
-  fun h ↦ by simpa [h] using P.root_coroot_two i
+variable (hp : ∀ i, p.toLin (root i) (coroot i) = 2)
 
-@[simp]
-lemma root_coroot_eq_pairing :
-    P.toLin (P.root i) (P.coroot j) = P.pairing i j :=
-  rfl
+private theorem choose_choose_eq_of_mapsTo :
+    (exist_eq_reflection_of_mapsTo p root coroot i
+      (exist_eq_reflection_of_mapsTo p root coroot i j h).choose h).choose = j := by
+  refine root.injective ?_
+  rw [(exist_eq_reflection_of_mapsTo p root coroot i _ h).choose_spec,
+    (exist_eq_reflection_of_mapsTo p root coroot i j h).choose_spec]
+  apply involutive_preReflection (x := root i) (hp i)
 
-lemma coroot_root_eq_pairing :
-    P.toLin.flip (P.coroot i) (P.root j) = P.pairing j i := by
-  simp
+/-- The bijection on the indexing set induced by reflection. -/
+@[simps]
+private def equiv_of_mapsTo :
```

Comment:
I'm not sure this should be `private`. Is `protected` enough?

Because this def is used in other defs that aren't private. So if those are unfolded, users are suddenly faced with a hideous name. Using a private def in the proof of a non-private lemma is fine.

In conclusion, I suggest:
```suggestion
protected def equiv_of_mapsTo :
```

Reply:
Good point, done.